packages/core/src/indexing/service.ts

```diff
@@ -527,264 +712,277 @@ export class IndexingService extends Emittery<IndexingEvents> {
+    }
+
+    // Update handled and matched events, unawaited because not in critical path.
+    this.setEventMetrics();
+
+    const abi = [keyHandler.abiEvent];
+    const contractName = keyHandler.contractName;
+    const eventName = keyHandler.eventName;
+
+    for (const event of events) {
+      try {
+        const decodedLog = decodeEventLog({
+          abi,
+          data: event.log.data,
+          topics: event.log.topics,
+        });
+
+        tasks.push({
+          kind: "LOG",
+          event: {
+            networkName: this.sourceById[event.sourceId].networkName,
+            contractName,
+            eventName,
+            event: {
+              args: decodedLog.args ?? {},
+              log: event.log,
+              block: event.block,
+              transaction: event.transaction,
+            },
+            checkpoint: {
+              blockNumber: Number(event.block.number),
+              blockTimestamp: Number(event.block.timestamp),
+              chainId: event.chainId,
+              logIndex: event.log.logIndex,
+            },
+          },
+        });
+      } catch (err) {
+        // Sometimes, logs match a selector but cannot be decoded using the provided ABI.
+        // This happens often when using custom event filters, because the indexed-ness
+        // of an event parameter is not taken into account when generating the selector.
+        this.common.logger.debug({
+          service: "app",
+          msg: `Unable to decode log, skipping it. id: ${event.log.id}, data: ${event.log.data}, topics: ${event.log.topics}`,
+        });
+      }
+    }
 
-    return queue;
+    // handle last event
+    if (tasks.length !== 0) {
+      tasks[tasks.length - 1].event.endCheckpoint =
+        keyHandler.maxTaskCheckpoint;
+
+      tasks[tasks.length - 1].event.eventsProcessed = events.length;
+    } else {
+      keyHandler.checkpoint = keyHandler.maxTaskCheckpoint;
+      this.emitCheckpoint();
+    }
   };
-}
 
-const buildContexts = (
-  sources: Source[],
-  networks: Network[],
-  requestQueues: RequestQueue[],
-  syncStore: SyncStore,
-  actions: ReturnType<typeof ponderActions>,
-) => {
-  const contexts: Record<
-    number,
-    {
-      client: Client;
-      network: { chainId: number; name: string };
-      contracts: Record<
-        string,
-        {
-          abi: Abi;
-          address?: Address | readonly Address[];
-          startBlock: number;
-          endBlock?: number;
-          maxBlockRange?: number;
-        }
-      >;
+  private emitCheckpoint = () => {
+    const checkpoint = checkpointMin(
+      ...Object.values(this.indexingFunctionMap!).map((i) => i.checkpoint),
+    );
+
+    this.emit("eventsProcessed", {
+      toCheckpoint: checkpoint,
+    });
+  };
+
+  private buildSourceById = () => {
+    for (const source of this.sources) {
+      this.sourceById[source.id] = source;
     }
-  > = {};
+  };
+
+  private buildIndexingFunctionMap = () => {
```

Comment:
I wonder if it's worth combining `this.indexingFunctions` and the top half of `this.indexingFunctionMap` into one `Record` that's keyed by event key (eg `"Contract:Event"`). This object could also be created in the build service.

Then, you could have another record keyed the same way calling smth like `indexingFunctionState`. This would enforce some degree of separation from the immutable and mutable pieces of state in the class which could be desirable. Low conviction comment though, take it or leave it.